worker/upgradedatabase/worker_test.go

```diff
@@ -0,0 +1,274 @@
+	"github.com/juju/juju/upgrades"
+	jujuversion "github.com/juju/juju/version"
+	"github.com/juju/juju/worker/upgradedatabase"
+	. "github.com/juju/juju/worker/upgradedatabase/mocks"
+)
+
+// baseSuite is embedded in both the worker and manifold tests.
+// Tests should not go on this suite directly.
+type baseSuite struct {
+	testing.IsolationSuite
+
+	logger *MockLogger
+}
+
+// ignoreLogging turns the suite's mock logger into a sink, with no validation.
+// Logs are still emitted via the test logger.
+func (s *baseSuite) ignoreLogging(c *gc.C) {
+	debugIt := func(message string, args ...interface{}) { logIt(c, loggo.DEBUG, message, args) }
+	infoIt := func(message string, args ...interface{}) { logIt(c, loggo.INFO, message, args) }
+	errorIt := func(message string, args ...interface{}) { logIt(c, loggo.ERROR, message, args) }
+
+	e := s.logger.EXPECT()
+	e.Debugf(gomock.Any(), gomock.Any()).AnyTimes().Do(debugIt)
+	e.Infof(gomock.Any(), gomock.Any()).AnyTimes().Do(infoIt)
+	e.Errorf(gomock.Any(), gomock.Any()).AnyTimes().Do(errorIt)
+}
+
+func logIt(c *gc.C, level loggo.Level, message string, args interface{}) {
+	var nArgs []interface{}
+	var ok bool
+	if nArgs, ok = args.([]interface{}); ok {
+		nArgs = append([]interface{}{level}, nArgs...)
+	} else {
+		nArgs = append([]interface{}{level}, args)
+	}
+
+	c.Logf("%s "+message, nArgs...)
+}
+
+type workerSuite struct {
+	baseSuite
+
+	lock      *MockLock
+	agent     *MockAgent
+	agentCfg  *MockConfig
+	cfgSetter *MockConfigSetter
+	pool      *MockPool
+}
+
+var _ = gc.Suite(&workerSuite{})
+
+func (s *workerSuite) TestValidateConfig(c *gc.C) {
+	defer s.setupMocks(c).Finish()
+
+	cfg := s.getConfig()
+	c.Check(cfg.Validate(), jc.ErrorIsNil)
+
+	cfg.UpgradeComplete = nil
+	c.Check(cfg.Validate(), jc.Satisfies, errors.IsNotValid)
+
+	cfg = s.getConfig()
+	cfg.Tag = nil
+	c.Check(cfg.Validate(), jc.Satisfies, errors.IsNotValid)
+
+	cfg = s.getConfig()
+	cfg.Agent = nil
+	c.Check(cfg.Validate(), jc.Satisfies, errors.IsNotValid)
+
+	cfg = s.getConfig()
+	cfg.Logger = nil
+	c.Check(cfg.Validate(), jc.Satisfies, errors.IsNotValid)
+
+	cfg = s.getConfig()
+	cfg.OpenState = nil
+	c.Check(cfg.Validate(), jc.Satisfies, errors.IsNotValid)
+
+	cfg = s.getConfig()
+	cfg.PerformUpgrade = nil
+	c.Check(cfg.Validate(), jc.Satisfies, errors.IsNotValid)
+
+	cfg = s.getConfig()
+	cfg.RetryStrategy = utils.AttemptStrategy{}
+	c.Check(cfg.Validate(), jc.Satisfies, errors.IsNotValid)
+}
+
+func (s *workerSuite) TestAlreadyCompleteNoWork(c *gc.C) {
+	defer s.setupMocks(c).Finish()
+	s.ignoreLogging(c)
+
+	s.lock.EXPECT().IsUnlocked().Return(true)
+
+	w, err := upgradedatabase.NewWorker(s.getConfig())
+	c.Assert(err, jc.ErrorIsNil)
+
+	workertest.CleanKill(c, w)
+}
+
+func (s *workerSuite) TestNotPrimaryNoWork(c *gc.C) {
+	defer s.setupMocks(c).Finish()
+	s.ignoreLogging(c)
+
+	s.lock.EXPECT().IsUnlocked().Return(false)
+	s.pool.EXPECT().IsPrimary("0").Return(false, nil)
+	s.lock.EXPECT().Unlock()
```

Comment:
I tend not to use `gomock.InOrder` unless the sequencing is logically required to orchestrate the tested behaviour. Otherwise it adds extra syntax and brittleness.